File: legacy-python/seed.py
"""Seed 8 realistic dummy FNOLs with synthetic photos + docs, then run analysis.

Usage:
    source .venv/bin/activate
    python -m uvicorn app:app --port 8765  # in another terminal
    python seed.py

Directly inserts claims and documents into the DB (bypasses the upload
endpoint's real LLM extraction — instead we set the extracted_json ourselves
so the rules engine and fraud detector can be exercised offline).
Then calls POST /claims/{id}/analyze on the running server so the full
aggregation + rules + fraud path runs end-to-end.
"""
import hashlib
import io
import json
import urllib.request
from datetime import date, timedelta
from pathlib import Path

from PIL import Image, ImageDraw
import imagehash

import db

BASE = Path(__file__).parent
UPLOADS = BASE / "uploads"
UPLOADS.mkdir(exist_ok=True)
BASE_URL = "http://127.0.0.1:8765"


def render_photo(claim_dir: Path, name: str, palette: tuple[int, int, int],
                 seed: int) -> tuple[bytes, str]:
    """Build a tiny synthetic damage photo, save to disk, return (bytes, path)."""
    img = Image.new("RGB", (320, 240), palette)
    d = ImageDraw.Draw(img)
    # deterministic "dent" pattern by seed
    for i in range(6):
        x = 40 + (seed * 17 + i * 31) % 220
        y = 40 + (seed * 13 + i * 23) % 140
        d.rectangle([x, y, x + 30, y + 20], fill=(20, 20, 20))
    d.text((10, 10), name, fill=(255, 255, 255))
    buf = io.BytesIO()
    img.save(buf, format="JPEG", quality=85)
    raw = buf.getvalue()
    path = claim_dir / name
    path.write_bytes(raw)
    return raw, str(path.relative_to(BASE))


def render_pdf_text(claim_dir: Path, name: str, body: str) -> tuple[bytes, str]:
    """Minimal 'PDF-ish' file: plain text with .pdf extension. The seed writes
    extracted_json directly so we don't need a real parser."""
    raw = body.encode()
    path = claim_dir / name
    path.write_bytes(raw)
    return raw, str(path.relative_to(BASE))


def add_photo(cid: str, claim_dir: Path, name: str, seed: int,
              extracted: dict, palette=(90, 110, 140),
              shared_bytes: bytes | None = None):
    if shared_bytes is not None:
        raw = shared_bytes
        path = claim_dir / name
        path.write_bytes(raw)
        filepath = str(path.relative_to(BASE))
    else:
        raw, filepath = render_photo(claim_dir, name, palette, seed)
    ch = hashlib.sha256(raw).hexdigest()
    with Image.open(io.BytesIO(raw)) as im:
        phash = str(imagehash.phash(im))
    db.add_document(
        claim_id=cid, filename=name, filepath=filepath, doc_type="photo",
        content_hash=ch, perceptual_hash=phash, extracted=extracted,
    )
    return raw


def add_estimate(cid: str, claim_dir: Path, amount: float, confidence: float):
    body = f"Repair estimate — parts + labour: EUR {amount:.2f}\n"
    raw, filepath = render_pdf_text(claim_dir, "repair_estimate.pdf", body)
    ch = hashlib.sha256(raw).hexdigest()
    ex = {
        "estimated_amount_eur": {"value": amount, "confidence": confidence},
        "labour_hours": {"value": 4.0, "confidence": confidence},
        "overall_confidence": confidence,
    }
    db.add_document(
        claim_id=cid, filename="repair_estimate.pdf", filepath=filepath,
        doc_type="repair_estimate", content_hash=ch, perceptual_hash=None,
        extracted=ex,
    )


def add_police_report(cid: str, claim_dir: Path, number: str):
    body = f"Politie proces-verbaal {number}\nBetrokken voertuigen: 2\n"
    raw, filepath = render_pdf_text(claim_dir, "police_report.pdf", body)
    ch = hashlib.sha256(raw).hexdigest()
    db.add_document(
        claim_id=cid, filename="police_report.pdf", filepath=filepath,
        doc_type="police_report", content_hash=ch, perceptual_hash=None,
        extracted={"police_report_number": {"value": number, "confidence": 0.95},
                   "overall_confidence": 0.9},
    )


def add_email(cid: str, claim_dir: Path, subject: str, body: str):
    text = f"Subject: {subject}\n\n{body}"
    raw, filepath = render_pdf_text(claim_dir, "customer_email.txt", text)
    ch = hashlib.sha256(raw).hexdigest()
    db.add_document(
        claim_id=cid, filename="customer_email.txt", filepath=filepath,
        doc_type="email", content_hash=ch, perceptual_hash=None,
        extracted={"overall_confidence": 0.7},
    )


def analyze(cid: str) -> None:
    req = urllib.request.Request(f"{BASE_URL}/claims/{cid}/analyze", method="POST")
    try:
        with urllib.request.urlopen(req) as resp:
            _ = resp.read()
    except urllib.error.HTTPError as e:
        # 303 redirect is expected
        if e.code != 303:
            raise


def d(offset_days: int) -> str:
    return (date.today() - timedelta(days=offset_days)).isoformat()


def seed_all():
    scenarios = []

    # 1 — Auto-approve: minor scratch, timely, docs complete.
    scenarios.append({
        "fnol": dict(policyholder_name="Jan de Vries", policy_number="NL-1001",
                     license_plate="12-ABC-3", vin="WVWZZZ1KZAW123456",
                     loss_date=d(2), loss_location="Amsterdam Zuidas",
                     description="Minor rear bumper scratch parking lot.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 1, "palette": (70, 90, 120),
        "damage_areas": ["rear_bumper"], "severity": "minor",
        "estimate": 850, "conf": 0.92, "police": False, "shared_photo": True,
    })

    # 2 — Auto-approve: fender damage, higher amount but below cap.
    scenarios.append({
        "fnol": dict(policyholder_name="Anna Bakker", policy_number="NL-1002",
                     license_plate="34-DEF-5", vin="",
                     loss_date=d(5), loss_location="Utrecht Centrum",
                     description="Fender scraped on bollard.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 2, "palette": (100, 100, 130),
        "damage_areas": ["front_fender"], "severity": "minor",
        "estimate": 1200, "conf": 0.89, "police": False,
    })

    # 3 — Assisted: third-party involved, own damage moderate.
    scenarios.append({
        "fnol": dict(policyholder_name="Ruben Peters", policy_number="NL-1003",
                     license_plate="56-GHI-7", vin="",
                     loss_date=d(3), loss_location="Rotterdam Kralingen",
                     description="Side-swipe by other vehicle at intersection.",
                     third_party_involved=True, injuries=False,
                     police_report_number="PV-2026-88123"),
        "photo_seed": 3, "palette": (90, 120, 100),
        "damage_areas": ["driver_door", "front_fender"], "severity": "moderate",
        "estimate": 1950, "conf": 0.88, "police": True,
    })

    # 4 — Manual: injury reported (hard blocker).
    scenarios.append({
        "fnol": dict(policyholder_name="Sofie Jansen", policy_number="NL-1004",
                     license_plate="78-JKL-9", vin="",
                     loss_date=d(1), loss_location="Den Haag Bezuidenhout",
                     description="T-bone collision, passenger reported neck pain.",
                     third_party_involved=True, injuries=True,
                     police_report_number="PV-2026-88410"),
        "photo_seed": 4, "palette": (140, 90, 90),
        "damage_areas": ["passenger_door", "b_pillar", "rear_quarter"],
        "severity": "severe", "estimate": 4400, "conf": 0.9, "police": True,
    })

    # 5 — Manual: over the €2500 cap.
    scenarios.append({
        "fnol": dict(policyholder_name="Marc de Boer", policy_number="NL-1005",
                     license_plate="90-MNO-1", vin="WBAAA1305C1234567",
                     loss_date=d(7), loss_location="Eindhoven",
                     description="Front-end collision with concrete pillar.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 5, "palette": (60, 60, 90),
        "damage_areas": ["hood", "front_bumper", "grille", "headlight_left"],
        "severity": "severe", "estimate": 5800, "conf": 0.91, "police": False,
    })

    # 6 — Manual: recycled photo (shares perceptual hash with #1).
    scenarios.append({
        "fnol": dict(policyholder_name="Piet van Dijk", policy_number="NL-1006",
                     license_plate="11-PQR-2", vin="",
                     loss_date=d(4), loss_location="Groningen",
                     description="Rear damage — see attached photo.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 1, "palette": (70, 90, 120),  # SAME as #1 => same phash
        "damage_areas": ["rear_bumper"], "severity": "minor",
        "estimate": 1100, "conf": 0.85, "police": False, "share_from": 0,
    })

    # 7 — Assisted: reported very late (>60d, <180d).
    scenarios.append({
        "fnol": dict(policyholder_name="Emma Visser", policy_number="NL-1007",
                     license_plate="22-STU-3", vin="",
                     loss_date=d(120), loss_location="Nijmegen",
                     description="Hail dent noticed after long trip.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 7, "palette": (110, 110, 110),
        "damage_areas": ["roof", "hood"], "severity": "moderate",
        "estimate": 1600, "conf": 0.86, "police": False,
    })

    # 8 — Manual: total loss classified.
    scenarios.append({
        "fnol": dict(policyholder_name="Lars Smit", policy_number="NL-1008",
                     license_plate="44-VWX-5", vin="",
                     loss_date=d(3), loss_location="Tilburg",
                     description="Vehicle deemed total loss after collision.",
                     third_party_involved=True, injuries=False,
                     police_report_number="PV-2026-90001"),
        "photo_seed": 8, "palette": (50, 40, 40),
        "damage_areas": ["hood", "front_bumper", "engine_bay", "windshield",
                         "roof", "a_pillar"],
        "severity": "total_loss", "estimate": 12500, "conf": 0.93, "police": True,
    })

    # 9 — Auto-approve: clean simple parking dent, everything filed cleanly.
    scenarios.append({
        "fnol": dict(policyholder_name="Femke Aarts", policy_number="NL-1009",
                     license_plate="55-YZA-6", vin="",
                     loss_date=d(1), loss_location="Haarlem",
                     description="Small parking dent from shopping cart.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 9, "palette": (95, 105, 125),
        "damage_areas": ["driver_door"], "severity": "minor",
        "estimate": 620, "conf": 0.94, "police": False,
    })

    # 10 — Manual + fraud routing: reported 260d late AND no supporting docs.
    #      Combined fraud (0.2 late + 0.15 no evidence + 0.05 photo-noexif noise) ≥ 0.30 → fraud gate blocks.
    scenarios.append({
        "fnol": dict(policyholder_name="Kees Nijhof", policy_number="NL-1010",
                     license_plate="66-BCD-7", vin="",
                     loss_date=d(260), loss_location="Enschede",
                     description="Just noticed old damage, submitting now.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 10, "palette": (130, 60, 60),
        "damage_areas": [], "severity": "unknown",
        "estimate": 900, "conf": 0.4,
        "no_documents": True,  # skip photo + estimate to trigger NO_EVIDENCE
    })

    # 11 — Assisted: customer email but no photos (soft evidence gap only).
    scenarios.append({
        "fnol": dict(policyholder_name="Yasmin Ozturk", policy_number="NL-1011",
                     license_plate="77-EFG-8", vin="",
                     loss_date=d(4), loss_location="Almere",
                     description="Front bumper crack, waiting on repair quote.",
                     third_party_involved=False, injuries=False),
        "photo_seed": 11, "palette": (100, 130, 110),
        "damage_areas": ["front_bumper"], "severity": "minor",
        "estimate": 1050, "conf": 0.82, "police": False, "email_only": True,
    })

    # Pass 1: create + upload docs.
    created_ids: list[str] = []
    photo_bytes_by_index: dict[int, bytes] = {}

    for idx, s in enumerate(scenarios):
        cid = db.create_claim(s["fnol"])
        created_ids.append(cid)
        claim_dir = UPLOADS / cid
        claim_dir.mkdir(exist_ok=True)

        if s.get("no_documents"):
            # NO_EVIDENCE scenario — deliberately upload nothing.
            print(f"[{idx+1}] created {cid} — {s['fnol']['policyholder_name']} (no docs)")
            continue

        if s.get("email_only"):
            # Only a customer email — no photo, no estimate. Triggers EVIDENCE_MINIMUM (soft).
            add_email(cid, claim_dir,
                      subject=f"Claim for {s['fnol']['license_plate']}",
                      body=s['fnol']['description'])
            print(f"[{idx+1}] created {cid} — {s['fnol']['policyholder_name']} (email only)")
            continue

        # Photo
        vision_extracted = {
            "damage_areas": s["damage_areas"],
            "severity": s["severity"],
            "confidence": 0.9,
            "notes": "seeded synthetic damage photo",
        }
        share_idx = s.get("share_from")
        shared = photo_bytes_by_index.get(share_idx) if share_idx is not None else None
        raw = add_photo(cid, claim_dir, f"damage_{idx+1}.jpg",
                        seed=s["photo_seed"], extracted=vision_extracted,
                        palette=s["palette"], shared_bytes=shared)
        if s.get("shared_photo") or share_idx is None:
            photo_bytes_by_index[idx] = raw

        # Repair estimate
        add_estimate(cid, claim_dir, s["estimate"], s["conf"])
        # Optional police report
        if s.get("police"):
            add_police_report(cid, claim_dir,
                              s["fnol"].get("police_report_number") or "PV-UNKNOWN")
        # Customer email — always include
        add_email(cid, claim_dir,
                  subject=f"Motor claim for {s['fnol']['license_plate']}",
                  body=s['fnol']['description'])

        print(f"[{idx+1}] created {cid} — {s['fnol']['policyholder_name']}")

    # Pass 2: run analysis on the live server.
    for cid in created_ids:
        analyze(cid)
        c = db.get_claim(cid)
        print(f"    analyzed {c['claim_number']}: "
              f"{c['decision_outcome']} · €{c['estimated_amount_eur']} · "
              f"fraud {c['fraud_score']}")


if __name__ == "__main__":
    db.init_db()
    seed_all()
    print("\nseed done — open http://127.0.0.1:8765/")

File: src/JbAutoAi/Seed.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Formats.Jpeg;
using SixLabors.ImageSharp.Metadata.Profiles.Exif;
using SixLabors.ImageSharp.PixelFormats;

namespace JbAutoAi;

/// Deterministic demo data: twelve claims chosen to exercise every decision path
/// and every fraud signal. Documents are written straight to the DB with their
/// extraction payload already filled in, so the seed runs offline; the analysis
/// stage then runs for real, exactly as it would after a live upload.
public static class Seed
{
    record Scenario(
        string Name, string Policy, string Plate, string Vin, int LossDaysAgo, string Location,
        string Description, bool ThirdParty, bool Injuries, string? PoliceNumber,
        int PhotoSeed, byte[] Palette, string[] DamageAreas, string Severity,
        double Estimate, double Confidence, bool Police = false, bool NoDocuments = false,
        bool EmailOnly = false, int? SharePhotoFrom = null, int? ExifDaysAgo = null);

    static DateOnly Days(int ago) => DateOnly.FromDateTime(DateTime.UtcNow).AddDays(-ago);

    static readonly Scenario[] Scenarios =
    [
        new("Jan de Vries", "NL-1001", "12-ABC-3", "WVWZZZ1KZAW123456", 2, "Amsterdam Zuidas",
            "Lichte kras op de achterbumper op de parkeerplaats.", false, false, null,
            1, [70, 90, 120], ["rear_bumper"], "minor", 850, 0.92),

        new("Anna Bakker", "NL-1002", "34-DEF-5", "", 5, "Utrecht Centrum",
            "Spatbord geschaafd langs een paaltje.", false, false, null,
            2, [100, 100, 130], ["front_fender"], "minor", 1200, 0.89),

        new("Ruben Peters", "NL-1003", "56-GHI-7", "", 3, "Rotterdam Kralingen",
            "Zijdelingse aanrijding door ander voertuig op kruising; tegenpartij kwam van rechts.",
            true, false, "PV-2026-88123",
            3, [90, 120, 100], ["driver_door", "front_fender"], "moderate", 1950, 0.88, Police: true),

        new("Sofie Jansen", "NL-1004", "78-JKL-9", "", 1, "Den Haag Bezuidenhout",
            "Flankaanrijding, passagier meldt nekklachten.", true, true, "PV-2026-88410",
            4, [140, 90, 90], ["passenger_door", "b_pillar", "rear_quarter"], "severe", 4400, 0.90,
            Police: true),

        new("Marc de Boer", "NL-1005", "90-MNO-1", "WBAAA1305C1234567", 7, "Eindhoven",
            "Frontale botsing tegen betonnen paal.", false, false, null,
            5, [60, 60, 90], ["hood", "front_bumper", "grille", "headlight_left"], "severe", 5800, 0.91),

        new("Piet van Dijk", "NL-1006", "11-PQR-2", "", 4, "Groningen",
            "Achterschade — zie bijgevoegde foto.", false, false, null,
            1, [70, 90, 120], ["rear_bumper"], "minor", 1100, 0.85, SharePhotoFrom: 0),

        new("Emma Visser", "NL-1007", "22-STU-3", "", 120, "Nijmegen",
            "Hagelschade pas opgemerkt na een lange rit.", false, false, null,
            7, [110, 110, 110], ["roof", "hood"], "moderate", 1600, 0.86),

        new("Lars Smit", "NL-1008", "44-VWX-5", "", 3, "Tilburg",
            "Voertuig total loss verklaard na aanrijding.", true, false, "PV-2026-90001",
            8, [50, 40, 40],
            ["hood", "front_bumper", "engine_bay", "windshield", "roof", "a_pillar"],
            "total_loss", 12500, 0.93, Police: true),

        new("Femke Aarts", "NL-1009", "55-YZA-6", "", 1, "Haarlem",
            "Kleine parkeerdeuk door winkelwagen.", false, false, null,
            9, [95, 105, 125], ["driver_door"], "minor", 620, 0.94),

        new("Kees Nijhof", "NL-1010", "66-BCD-7", "", 260, "Enschede",
            "Oude schade nu pas gemeld.", false, false, null,
            10, [130, 60, 60], [], "unknown", 900, 0.40, NoDocuments: true),

        new("Yasmin Ozturk", "NL-1011", "77-EFG-8", "", 4, "Almere",
            "Scheur in de voorbumper, offerte volgt nog.", false, false, null,
            11, [100, 130, 110], ["front_bumper"], "minor", 1050, 0.82, EmailOnly: true),

        // Fietser + EXIF-mismatch: art. 185 WVW hard gate and a forensics signal in one.
        new("Tom Aalders", "NL-1012", "88-HIJ-9", "", 6, "Leiden Breestraat",
            "Aanrijding met een fietser die van rechts kwam; fietser is 12 jaar oud.",
            true, true, "PV-2026-91188",
            12, [80, 100, 90], ["front_bumper", "headlight_right"], "moderate", 2100, 0.87,
            Police: true, ExifDaysAgo: 400),
    ];

    /// The demo policyholder account. Owns the first three seeded claims, so the
    /// portal has something to show on first login.
    public const string DemoPortalEmail = "jan.devries@example.nl";
    const int DemoPortalClaimCount = 3;

    static string EnsureDemoPortalUser()
    {
        var existing = Db.GetPortalUserByEmail(DemoPortalEmail);
        if (existing is not null) return existing.Id;
        var created = Db.CreatePortalUser(DemoPortalEmail, "Jan de Vries", "+31 6 1234 5678",
                                          Auth.HashPassword(Db.DemoPassword));
        return created?.Id ?? Db.GetPortalUserByEmail(DemoPortalEmail)!.Id;
    }

    /// (palette, seed) of every fixture photo, in scenario order. Smoke asserts
    /// these stay perceptually distinct — see Smoke.MediaChecks.
    internal static IEnumerable<(byte[] Palette, int Seed, int Index)> PhotoFixtures() =>
        Scenarios.Select((s, i) => (s.Palette, s.PhotoSeed, i))
                 .Where(t => !Scenarios[t.i].NoDocuments && !Scenarios[t.i].EmailOnly);

    public static async Task RunAsync()
    {
        Directory.CreateDirectory(Pipeline.UploadsRoot);
        var portalUserId = EnsureDemoPortalUser();
        var created = new List<string>();
        var photoBytesByIndex = new Dictionary<int, byte[]>();

        for (var i = 0; i < Scenarios.Length; i++)
        {
            var s = Scenarios[i];
            var claimId = Db.CreateClaim(new Claim
            {
                PolicyholderName = s.Name,
                PolicyNumber = s.Policy,
                LicensePlate = s.Plate,
                Vin = s.Vin,
                LossDate = Days(s.LossDaysAgo),
                LossLocation = s.Location,
                Description = s.Description,
                ThirdPartyInvolved = s.ThirdParty,
                Injuries = s.Injuries,
                PoliceReportNumber = s.PoliceNumber,
            });
            created.Add(claimId);

            // Give the demo portal account a few claims to look at.
            if (i < DemoPortalClaimCount)
            {
                Db.LinkClaimToPortalUser(claimId, portalUserId);
                Db.AddCustomerActivity(claimId, "created", portalUserId, null);
            }

            var dir = Path.Combine(Pipeline.UploadsRoot, claimId);
            Directory.CreateDirectory(dir);

            if (s.NoDocuments)
            {
                Console.WriteLine($"[{i + 1,2}] {claimId} — {s.Name} (no documents)");
                continue;
            }

            if (s.EmailOnly)
            {
                AddEmail(claimId, dir, s);
                Console.WriteLine($"[{i + 1,2}] {claimId} — {s.Name} (email only)");
                continue;
            }

            var shared = s.SharePhotoFrom is { } from ? photoBytesByIndex.GetValueOrDefault(from) : null;
            var raw = AddPhoto(claimId, dir, $"damage_{i + 1}.jpg", s, shared);
            photoBytesByIndex[i] = raw;

            AddEstimate(claimId, dir, s);
            if (s.Police) AddPoliceReport(claimId, dir, s);
            AddEmail(claimId, dir, s);

            Console.WriteLine($"[{i + 1,2}] {claimId} — {s.Name}");
        }

        Console.WriteLine("\nrunning analysis…\n");
        foreach (var id in created)
        {
            await Pipeline.AnalyseAsync(id);
            var c = Db.GetClaim(id)!;
            Console.WriteLine($"  {c.ClaimNumber}  {c.DecisionOutcome,-14} "
                            + $"€{c.EstimatedAmountEur,8:F0}  fraud {c.FraudScore:F2}  "
                            + $"citations {c.CitationIntegrity:P0}");
        }

        var health = Db.GetCitationHealth();
        Console.WriteLine($"\nseed done — {created.Count} claims, corpus {health.CorpusVersion}, "
                        + $"{health.CorpusEmbedded}/{health.CorpusChunks} chunks embedded, "
                        + $"citation integrity {health.Integrity:P0}.");
    }

    /// Deterministic synthetic damage photo — same seed produces byte-identical
    /// output, which is what makes the recycled-photo scenario reproducible.
    /// Deterministic synthetic damage photo.
    ///
    /// The variation has to live in LOW spatial frequencies. A DCT perceptual hash
    /// keeps only the top-left 8×8 coefficients, so per-pixel noise — which is where
    /// the obvious approach puts the entropy — is discarded wholesale and unrelated
    /// fixtures land a handful of bits apart, false-positiving the recycled-photo
    /// check. Real photos differ in their broad structure; so must these.
    internal static byte[] RenderPhoto(byte[] palette, int seed, DateOnly? exifDate = null)
    {
        const int w = 320, h = 240;
        using var img = new Image<Rgb24>(w, h);

        var rng = new Random(seed * 7919);
        // Per-seed low-frequency field: a tilted gradient plus two sinusoids whose
        // orientation and period come from the seed.
        double gx = (seed * 37 % 100) / 100.0 - 0.5, gy = (seed * 61 % 100) / 100.0 - 0.5;
        double f1 = 1 + seed % 4, f2 = 1 + (seed * 3) % 5, phase = seed * 0.7;

        for (var y = 0; y < h; y++)
            for (var x = 0; x < w; x++)
            {
                var u = (double)x / w;
                var v = (double)y / h;
                var field = 70 * (gx * u + gy * v)
                          + 34 * Math.Sin(Math.PI * f1 * u + phase)
                          + 28 * Math.Cos(Math.PI * f2 * v + phase)
                          + rng.Next(-10, 11);
                img[x, y] = new Rgb24(
                    (byte)Math.Clamp(palette[0] + field, 0, 255),
                    (byte)Math.Clamp(palette[1] + field, 0, 255),
                    (byte)Math.Clamp(palette[2] + field, 0, 255));
            }

        // Dark "damage" blocks — count, size and position all seed-dependent.
        var blocks = 3 + seed % 6;
        for (var i = 0; i < blocks; i++)
        {
            var bw = 26 + (seed * 11 + i * 17) % 60;
            var bh = 18 + (seed * 7 + i * 13) % 44;
            var x0 = (seed * 17 + i * 53) % (w - bw);
            var y0 = (seed * 13 + i * 41) % (h - bh);
            for (var y = y0; y < y0 + bh; y++)
                for (var x = x0; x < x0 + bw; x++)
                    img[x, y] = new Rgb24(18, 18, 20);
        }

        if (exifDate is { } taken)
        {
            var profile = new ExifProfile();
            profile.SetValue(ExifTag.DateTimeOriginal,
                taken.ToDateTime(TimeOnly.MinValue).ToString("yyyy:MM:dd HH:mm:ss"));
            img.Metadata.ExifProfile = profile;
        }

        using var ms = new MemoryStream();
        img.Save(ms, new JpegEncoder { Quality = 85 });
        return ms.ToArray();
    }

    static byte[] AddPhoto(string claimId, string dir, string name, Scenario s, byte[]? shared)
    {
        var raw = shared ?? RenderPhoto(s.Palette, s.PhotoSeed,
                                        s.ExifDaysAgo is { } d ? Days(d) : null);
        var path = Path.Combine(dir, name);
        File.WriteAllBytes(path, raw);

        var extracted = Json.Str(new Dictionary<string, object?>
        {
            ["damage_areas"] = s.DamageAreas,
            ["severity"] = s.Severity,
            ["confidence"] = 0.9,
            ["photo_quality"] = "good",
            ["notes"] = "seeded synthetic damage photo",
            ["photo_signals"] = Media.PhotoExifSignals(raw, Days(s.LossDaysAgo)),
        });

        Db.AddDocument(claimId, name, $"{claimId}/{name}", "photo",
                       Media.Sha256Hex(raw), Media.PerceptualHash(raw), extracted);
        return raw;
    }

    static void AddEstimate(string claimId, string dir, Scenario s)
    {
        var body = $"Reparatiecalculatie — onderdelen + arbeidsloon: EUR {s.Estimate:F2}\n";
        var raw = WriteFile(dir, "repair_estimate.pdf", body);
        var extracted = Json.Str(new Dictionary<string, object?>
        {
            ["estimated_amount_eur"] = new { value = s.Estimate, confidence = s.Confidence },
            ["labour_hours"] = new { value = 4.0, confidence = s.Confidence },
            ["overall_confidence"] = s.Confidence,
        });
        Db.AddDocument(claimId, "repair_estimate.pdf", $"{claimId}/repair_estimate.pdf",
                       "repair_estimate", Media.Sha256Hex(raw), null, extracted);
    }

    static void AddPoliceReport(string claimId, string dir, Scenario s)
    {
        var number = s.PoliceNumber ?? "PV-ONBEKEND";
        var raw = WriteFile(dir, "police_report.pdf",
            $"Politie proces-verbaal {number}\nBetrokken voertuigen: 2\n{s.Description}\n");
        var extracted = Json.Str(new Dictionary<string, object?>
        {
            ["police_report_number"] = new { value = number, confidence = 0.95 },
            ["overall_confidence"] = 0.9,
        });
        Db.AddDocument(claimId, "police_report.pdf", $"{claimId}/police_report.pdf",
                       "police_report", Media.Sha256Hex(raw), null, extracted);
    }

    static void AddEmail(string claimId, string dir, Scenario s)
    {
        var raw = WriteFile(dir, "customer_email.txt",
            $"Onderwerp: Schademelding {s.Plate}\n\n{s.Description}\n");
        Db.AddDocument(claimId, "customer_email.txt", $"{claimId}/customer_email.txt",
                       "email", Media.Sha256Hex(raw), null,
                       Json.Str(new Dictionary<string, object?> { ["overall_confidence"] = 0.7 }));
    }

    static byte[] WriteFile(string dir, string name, string body)
    {
        var raw = System.Text.Encoding.UTF8.GetBytes(body);
        File.WriteAllBytes(Path.Combine(dir, name), raw);
        return raw;
    }
}
